cloudify_hostpool/rest/backend.py

```diff
@@ -1,164 +1,370 @@
-# time, and all other instances will use the previously loaded pool.
-INDICATOR = 'host-pool-loaded-indicator'
 
+class HostAlchemist(object):
+    '''Converts user-provided host entries into a consumable structure'''
+    def __init__(self, config):
+        self.config = config
 
-class RestBackend(object):
+    def parse(self):
+        '''Performs the config-to-hosts analysis and conversion'''
+        hosts = []
+        defaults = self.get_config_defaults()
+        # Fix defaults
+        if defaults.get('platform'):
+            del defaults['platform']
+        if defaults.get('endpoint', dict()).get('ip'):
+            del defaults['endpoint']['ip']
+        # Validate the defaults
+        self.validate_defaults(defaults)
+        # Build an extended list of hosts
+        for host in self.get_config_hosts():
+            # Merge defaults into hosts
+            self.impose_defaults(host, defaults)
+            # Soft-validate the base host
+            self.validate_host(host, check_ip_range=False)
+            hip = IPNetwork(host['endpoint']['ip'])
+            hname = host['name']
+            if len(list(hip)) > 1:
+                # This is an IP address range endpoint
+                for idx, sip in enumerate(list(hip)):
+                    shost = deepcopy(host)
+                    shost['endpoint']['ip'] = str(sip)
+                    if hname:
+                        shost['name'] = '{0}_{1}'.format(hname, idx)
+                    hosts.append(shost)
+            else:
+                # This is a single IP endpoint
+                host['endpoint']['ip'] = str(list(hip)[0])
+                hosts.append(host)
+        # Post-process hosts
+        for host in hosts:
+            # Add in backend details
+            host['allocated'] = False
+            host['alive'] = False
+            # Validate the extended hosts list
+            self.validate_host(host)
+        return hosts
 
-    def __init__(self, pool, storage=None):
-        self.storage = sqlite.SQLiteStorage(storage)
-        # allow only one process to do the initial load
+    def get_config_defaults(self):
+        '''Returns the default config'''
+        return self.config.get('default', {})
 
-        def _create_indicator():
-            fd = os.open(INDICATOR, os.O_WRONLY |
-                         os.O_CREAT | os.O_EXCL, 0600)
-            os.close(fd)
+    def get_config_hosts(self):
+        '''Returns the hosts config'''
+        return self.config.get('hosts', [])
 
-        with FLock:
-            if not os.path.exists(INDICATOR):
-                self._load_pool(pool)
-                _create_indicator()
+    @staticmethod
+    def validate_defaults(defaults):
+        '''Validates default data'''
+        if not defaults:
+            return
+        if not isinstance(defaults, dict):
+            raise exceptions.ConfigurationError(
+                'Defaults must be a valid JSON Object')
+        if defaults.get('platform'):
```

Comment:
True, if you are to assume that TestBackend::parse() is the only entry point that will ever be used.  I'd assume someone, at some point, will directly use the method and we want that check in place.